rgmanager/man/clusvcadm.8
.TH "clusvcadm" "8" "Jan 2005" "" "Red Hat Cluster Suite"
.SH "NAME"
clusvcadm \- Cluster User Service Administration Utility
.SH "SYNOPSIS"
.B clusvcadm
.B [\-d
.I <service>
.B ]
.B [\-e
.I <service>
.B ]
.B [\-l]
.B [\-u]
.B [\-S]
.B [\-m
.I <member>
.B ]
.B [\-r
.I <service>
.B ]
.B [\-R
.I <service>
.B ]
.B [\-M
.I <service>
.B ]
.B [\-Z
.I <service>
.B ]
.B [\-U
.I <service>
.B ]
.B [\-s
.I <service>
.B ]
.B [\-c
.I <service>
.B ]
.B [\-v]

.SH "DESCRIPTION"
.PP 
The
.B clusvcadm
command allows an administrator to enable, disable, relocate, and restart
user services in a cluster.  In order to perform cluster service operations,
the cluster daemons must be running (and have quorum) on the member system
on which the command is invoked.

.SH "OPTIONS"
.IP "\-d <service>"
Stops and disables the user service named
.I
service
.IP "\-e <service>"
Enables and starts the user service named
.I
service
.IP \-l
Lock services in preparation for cluster shutdown.  This should only
be used if the administrator intends to perform a global, cluster
wide shutdown.  This prevents services from starting (but not stopping,
like
.B -Z
does).  Once the cluster quorum is dissolved, this state is reset.
If a new instance of rgmanager boots while others are locked, the
behavior is undefined.
.IP \-u
Unlock resource group managers.  This allows services to start again. 
.IP "\-S"
Display whether the resource group managers are locked or not.  This
can be used to verify the correct operation of the \fB-l\fR and \fB-u\fR 
options, but is only useful for debugging.
.IP "\-m <member>"
When used in conjunction with either the
.B
\-e
or
.B
\-r
options, this specifies the 
.I
preferred
target member on which to start the 
service.
.IP "\-r <service>"
Relocates the user service named
.I
service
to another cluster member.
.IP "\-R <service>"
Restarts the user service named
.I
service
on the cluster member on which it is currently running.
.IP "\-M <service>"
Use a special "migration" operation to move the user service named
.I
service
to another cluster member.  Currently, this is only useful for
virtual machines.  Use of migration usually requires special
configuration of the local virtual machine manager in order
to work correctly.
.IP "\-Z <service>"
Freezes the service named
.I
service
on the cluster member on which it is currently running.  This will
prevent status checks of the service as well as failover in the
event the node fails or rgmanager is stopped.
.IP "\-U <service>"
Unfreezes the user service named
.I
service
on the cluster member on which it is currently running.  This will
re-enable status checks.
.IP "\-s <service>"
Stops the service named
.I
service
until a member transition or until it is enabled again.

.IP "\-c <service>"
Attempt to convalesce the named
.I
service
by restarting failed, non-critical components.

.IP \-v
Display version information and exit.

.SH "NOTES"
Executing
.I -U
(unfreeze) on a service which was frozen in the 
.B started
state while the service owner is offline results in an undefined
(and possibly dangerous) condition.  Manually ensure all resources are
clear before doing this.

.SH "SEE ALSO"
clustat(8)


rgmanager/ChangeLog
2008-04-16 Fabio M. Di Nitto <fabbione at fabbione.net>
	* src/utils/clustat.c: Fix variable init

2008-03-04 Lon Hohberger <lhh at redhat.com>
	* src/resources/ASEHAagent.sh, Makefile: Add Sybase failover agent
	* src/resources/oracledb.sh, Makefile: Add Oracle 10g failover agent

2008-02-26 Lon Hohberger <lhh at redhat.com>
	* src/resources/ip.sh: Fix netmask handling in ip.sh
	* src/utils/clustat.c: Don't show estranged nodes if they're down
2008-02-01 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_state.c, slang_event.c, groups.c, restart_counter.c,
	include/restart_counter.h, resgroup.h,
	src/resources/default_event_script.sl: Allow restart counters to 
	correctly work with central_processing. (#400211 / #431130)

2008-01-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_thread.c: Fix case that broke 'clusvcadm -e <service>
	-n <node>' case #430220
	* src/daemons/rg_state.c: Don't enable migrate-to-self (#430272)

2008-01-02 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/resources/ocf-shellfuncs: Add interpreter.

2007-12-30 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/daemons/rg_state.c, include/resgroup.h, src/clulib/alloc.c:
	  Fix building when -DDEBUG is defined.

2007-12-24 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/clulib/vft.c: Change ifdef to fix build on parisc.

2007-12-19 Lon Hohberger <lhh at redhat.com>
	* include/res-ocf.h: Fix typo
	* src/daemons/restree.c: Export reference counts to resource agents.
	* src/resources/clusterfs.sh: Use reference counts to ensure we
	don't unmount a GFS file system if it's in use by multiple services
	on the same node.  #254111
	* src/resources/service.sh, default_event_script.sl: Add support for
	"soft" dependencies if central_processing is enabled

2007-12-14 Lon Hohberger <lhh at redhat.com>
	(Misc central_events fixes)
	* Add return code for inability to run because of exclusive 
	tag
	* Only initialize state once
	* Fix master-node failover.
	* Correctly return node after enable if request was forwarded

2007-12-12 Lon Hohberger <lhh at redhat.com>
	* Misc changes; add missing ds.h
	* src/resources/default*.sl: Make clusvcadm -r go to a different
	node.  Don't notice() that we're starting a service if it's
	disabled.
	* src/daemons/groups.c: Make clustat not interfere with startup
	when central processing is enabled
	* src/daemons/slang_event.c: Make clusvcadm correctly report
	service owner after relocate/restart/enable requests.

2007-11-30 Lon Hohberger <lhh at redhat.com>
	* Commit RIND / S-Lang script engine [untested]

[RHEL5 merged ChangeLog Entries]
2007-11-30 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Retry mount up to 3 times to avoid
	race condition during another process mounting a GFS volume
	* src/resources/vm.sh, service.sh: Add defaults for values.
	Make vm.sh work with more service attrs (max restarts)
	* src/utils/clustat.c: Make output of clustat terminal-width 
	dependent

2007-11-26 Lon Hohberger <lhh at redhat.com>
	* include/reslist.h: Add restart counters to resource node structure
	(intended for top-level resources, i.e. services, vms...)
	* include/restart_counter.h: Add header file for restart counter
	* src/daemons/Makefile: Fix build to include restart counters
	* src/daemons/restart_counter.c: Implement restart counters #247139
	* src/daemons/fo_domain.c, groups.c, restart_counter.c, resrules.c,
	restree.c, test.c: Glue for restart counters.
	* src/daemons/reslist.c: Glue for restart counters.  Make expand_time
	parser more robust to allow things like '1h30m' as a time value.
	* src/daemons/main.c: Mark quorum disk offline in the correct
	place to avoid extraneous log messages
	* src/daemons/rg_state.c: Allow marking service as stopped if
	stuck in recover state.  Make service which failed to start
	go to stopped state.  Glue for restart counters.
	* src/resources/service.sh, vm.sh: Add parameters for restart
	counters #247139

2007-11-14 Lon Hohberger <lhh at redhat.com>
	* src/utils/clulog.c: Make clulog honor rgmanager log levels
	(#289501)
	* src/clulib/vft.c: Fix #303981 - crash on rgmanager restart in some
	cases
	* man/clusvcadm.8: Remove references to clushutdown from man page;
	resolves #324151
	* src/resources/netfs.sh: Apply patch from Marco Ceci to fix #358161
	* src/resources/vm.sh: Make default migration policy live instead
	of pause for Xen virtual machines.  Also make it configurable instead
	of static.  Resolves #345871

2007-11-13 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Add support for self_fence operation
	to clusterfs resource agent
	* src/resources/service.sh: Add default values to service.sh

2007-10-26 Lon Hohberger <lhh at redhat.com>
	* src/daemons/main.c, src/utils/clustat.c, clusvcadm.c:
	Fix #354391

2007-09-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c: Apply patch to fix side case re: 229650
	Patch from Simone Gotti. Resolves: #229650
[End RHEL5 merged changes]

2007-08-30 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c, rg_state.c: Fix tree-restart bug
	This is another part of #229650

2007-08-22 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* include/list.h: Fix build with gcc-4.2.

2007-08-15 Lon Hohberger <lhh at redhat.com>
	* src/utils/clustat.c: Remove uninitialized var & fix logic /
	spotted by Ryan McCabe

2007-08-09  Fabio M. Di Nitto <fabbione at ubuntu.com>
	* rgmanager/src/clulib/alloc.c: Undefine DEBUG when building on
	PARISC. __builtin_address needs the same love as IA64.

2007-08-02 Lon Hohberger <lhh at redhat.com>
	* general: More fixes around #248727
	* include/reslist.h, src/daemons/restree.c: Make last-value be
	returned or resources which have been checked recently
	* src/daemons/groups.c: Make VMs use migrate semantics instead of
	relocate semantics when employing failover domain rules
	* src/daemons/nodeevent.c: Fix VMs ending up on wrong nodes when
	simultaneous boot occurs
	* src/daemons/rg_forward.c: Fix erroneous timeout

rgmanager/src/utils/clusvcadm.c
/** @file
 * The New And Improved Cluster Service Admin Utility.
 * TODO Clean up the code.
 */
#include <stdio.h>
#include <stdlib.h>
#include <sys/param.h>
#include <libgen.h>
#include <resgroup.h>
#include <platform.h>
#include <members.h>
#include <message.h>
#include <libcman.h>
#include <resgroup.h>
#include <msgsimple.h>
#include <signal.h>

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


static void
build_message(SmMessageSt *msgp, int action, char *svcName, int target,
		int arg1, int arg2)
{
	msgp->sm_hdr.gh_magic = GENERIC_HDR_MAGIC;
	msgp->sm_hdr.gh_command = RG_ACTION_REQUEST;
	msgp->sm_hdr.gh_length = sizeof(*msgp);
	msgp->sm_hdr.gh_arg1 = arg1;
	msgp->sm_hdr.gh_arg2 = arg2;
	msgp->sm_data.d_action = action;
	strncpy(msgp->sm_data.d_svcName, svcName,
		sizeof(msgp->sm_data.d_svcName));
	msgp->sm_data.d_svcOwner = target;
	msgp->sm_data.d_ret = 0;

	swab_SmMessageSt(msgp);
}


static int
do_lock_req(int req)
{
	cman_handle_t ch;
	msgctx_t ctx;
	int ret = RG_FAIL;
	cluster_member_list_t *membership = NULL;
	int me;
	generic_msg_hdr hdr;

	ch = cman_init(NULL);
	if (!ch) {
		printf("Could not connect to cluster service\n");
		goto out;
	}

	membership = get_member_list(ch);
	me = get_my_nodeid(ch);

	if (msg_open(MSG_SOCKET, 0, RG_PORT, &ctx, 5) < 0) {
		printf("Could not connect to resource group manager\n");
		goto out;
	}

	if (msg_send_simple(&ctx, req, 0, 0) < 0) {
		printf("Communication failed\n");
		goto out;
	}

	if (msg_receive(&ctx, &hdr, sizeof(hdr), 5) < sizeof(hdr)) {
		printf("Receive failed\n");
		goto out;
	}

	swab_generic_msg_hdr(&hdr);
	ret = hdr.gh_command;

out:
	if (membership)
		free_member_list(membership);

	if (ctx.type >= 0)
		msg_close(&ctx);

	if (ch)
		cman_finish(ch);

	return ret;
}


static int
do_lock(void)
{
	if (do_lock_req(RG_LOCK) != RG_SUCCESS) {
		printf("Lock operation failed\n");
		return 1;
	}
	printf("Resource groups locked\n");
	return 0;
}


static int
do_unlock(void)
{
	if (do_lock_req(RG_UNLOCK) != RG_SUCCESS) {
		printf("Unlock operation failed\n");
		return 1;
	}
	printf("Resource groups unlocked\n");
	return 0;
}


static int
do_query_lock(void)
{
	switch(do_lock_req(RG_QUERY_LOCK)) {
	case RG_LOCK:
		printf("Resource groups locked\n");
		break;
	case RG_UNLOCK:
		printf("Resource groups unlocked\n");
		break;
	default:
		printf("Query operation failed\n");
		return 1;
	}
	return 0;
}


static void
usage(char *name)
{
printf("usage: %s [command]\n\n", name);

printf("Resource Group Control Commands:\n");
printf("  -v                     Display version and exit\n");
printf("  -d <group>             Disable <group>.  This stops a group\n"
       "                         until an administrator enables it again,\n"
       "                         the cluster loses and regains quorum, or\n"
       "                         an administrator-defined event script\n"
       "                         explicitly enables it again.\n");
printf("  -e <group>             Enable <group>\n");
printf("  -e <group> -F          Enable <group> according to failover\n"
       "                         domain rules (deprecated; always the\n"
       "                         case when using central processing)\n");
printf("  -e <group> -m <member> Enable <group> on <member>\n");
printf("  -r <group> -m <member> Relocate <group> [to <member>]\n"
       "                         Stops a group and starts it on another\n"
       "                         cluster member.\n");
printf("  -M <group> -m <member> Migrate <group> to <member>\n");
printf("                         (e.g. for live migration of VMs)\n");
printf("  -q                     Quiet operation\n");
printf("  -R <group>             Restart a group in place.\n");
printf("  -s <group>             Stop <group>.  This temporarily stops\n"
       "                         a group.  After the next group or\n"
       "                         or cluster member transition, the group\n"
       "                         will be restarted (if possible).\n");
printf("  -Z <group>             Freeze resource group.  This prevents\n"
       "                         transitions and status checks, and is \n"
       "                         useful if an administrator needs to \n"
       "                         administer part of a service without \n"
       "                         stopping the whole service.\n");
printf("  -U <group>             Unfreeze (thaw) resource group.  Restores\n"
       "                         a group to normal operation.\n");
printf("  -c <group>             Convalesce (repair, fix) resource group.\n"
       "                         Attempts to start failed, non-critical \n"
       "                         resources within a resource group.\n");

printf("Resource Group Locking (for cluster Shutdown / Debugging):\n");
printf("  -l                     Lock local resource group managers.\n"
       "                         This prevents resource groups from\n"
       "                         starting.\n");
printf("  -S                     Show lock state\n");
printf("  -u                     Unlock resource group managers.\n"
       "                         This allows resource groups to start.\n");
}


static int
find_closest_node(cluster_member_list_t *cml, char *name, size_t maxlen)
{
	int x, c = 0, cl = 0, nc = 0, ncl = 0, cur = 0;

	for (x=0; x<cml->cml_count; x++) {
		cur = 0;

		while (cml->cml_members[x].cn_name[cur] && name[cur] &&
		       (cml->cml_members[x].cn_name[cur] == name[cur]))
			cur++;
		if (!cur)
			continue;
		if (cur >= cl) {
			ncl = cl; /* Next-closest */
			nc = c;
			cl = cur;
			c = x;
		}
	}

	if (!cl) {
		printf("No matches for '%s' found\n", name);
		return 0;
	}

	if (ncl == cl) {
		printf("More than one possible match for '%s' found\n",
		       name);
		return 0;
	}

	printf("Closest match: '%s'\n", 
	       cml->cml_members[c].cn_name);

	strncpy(name, cml->cml_members[c].cn_name, maxlen);
	return cml->cml_members[c].cn_nodeid;
}


int
main(int argc, char **argv)
{
	extern char *optarg;
	char *svcname=NULL, nodename[256], realsvcname[64];
	int opt;
	msgctx_t ctx;
	cman_handle_t ch;
	SmMessageSt msg;
	generic_msg_hdr *h = (generic_msg_hdr *)&msg;
	int action = RG_STATUS;
	int fod = 0;
	int node_specified = 0;
       	int me, svctarget = 0;
	const char *actionstr = NULL;
	cluster_member_list_t *membership;

	while ((opt = getopt(argc, argv, "lSue:M:d:r:n:c:m:FvR:s:Z:U:qh?")) != EOF) {
		switch (opt) {
		case 'l':
			return do_lock();

		case 'S':
			return do_query_lock();

		case 'u':
			return do_unlock();

		case 'e':
			/* ENABLE */
			actionstr = "trying to enable";
			action = RG_ENABLE;
			svcname = optarg;
			break;
               case 'F':
                       if (node_specified) {
                               fprintf(stderr,
                                       "Cannot use '-F' with '-n' or '-m'\n");
                               return 1;
                       }
                       fod = 1;
                       break;
		case 'd':
			/* DISABLE */
			actionstr = "disabling";
			action = RG_DISABLE;
			svcname = optarg;
			break;
		case 'r':
			/* RELOCATE */
			actionstr = "relocate";
			action = RG_RELOCATE;
			svcname = optarg;
			break;
		case 'M':
			/* MIGRATE */
			actionstr = "migrate";
			action = RG_MIGRATE;
			svcname = optarg;
			break;
		case 's':
			/* stop */
			actionstr = "stopping";
			action = RG_STOP_USER;
			svcname = optarg;
			break;
		case 'R':
			actionstr = "trying to restart";
			action = RG_RESTART;
			svcname = optarg;
			break;
		case 'm': /* member ... */
		case 'n': /* node .. same thing */
                       if (fod) {
                               fprintf(stderr,
                                       "Cannot use '-F' with '-n' or '-m'\n");
                               return 1;
                       }
			strncpy(nodename,optarg,sizeof(nodename));
			node_specified = 1;
			break;
		case 'v':
			printf("%s\n", RELEASE_VERSION);
			return 0;
		case 'Z':
			actionstr = "freezing";
			action = RG_FREEZE;
			svcname = optarg;
			break;
		case 'U':
			actionstr = "unfreezing";
			action = RG_UNFREEZE;
			svcname = optarg;
			break;
		case 'c':
			actionstr = "convalescing";
			action = RG_CONVALESCE;
			svcname = optarg;
			break;
		case 'q':
			close(STDOUT_FILENO);
			break;
		case 'h':
		case '?':
		default:
			usage(basename(argv[0]));
			return 1;
		}
	}

	if (!svcname) {
		usage(basename(argv[0]));
		return 1;
	}

	if (action == RG_MIGRATE && !node_specified) {
		printf("Migration requires a target cluster member\n");
		return 1;
	}

	if (!strchr(svcname,':')) {
		snprintf(realsvcname, sizeof(realsvcname), "service:%s",
			 svcname);
		svcname = realsvcname;
	}

	signal(SIGPIPE, SIG_IGN);

	/* No login */
	ch = cman_init(NULL);
	if (!ch) {
		printf("Could not connect to cluster service\n");
		return 1;
	}

	membership = get_member_list(ch);
	me = get_my_nodeid(ch);

	if (node_specified) {
		svctarget = memb_name_to_id(membership, nodename);
		if (svctarget == 0) {
			printf("'%s' not in membership list\n",
			       nodename);
			svctarget = find_closest_node(membership, nodename,
						      sizeof(nodename));
			if (!svctarget)
				return 1;
		}
		if (action == RG_MIGRATE && 
		    memb_online(membership, svctarget) == 0) {
			printf("'%s' is offline\n", nodename);
			return 1;
		}
	} else {
		svctarget = 0;
		/*
		clu_local_nodename(RG_SERVICE_GROUP, nodename,
				   sizeof(nodename));
				   */
		//strcpy(nodename,"me");
	}

	build_message(&msg, action, svcname, svctarget, fod, 0);

	if (action != RG_RELOCATE && action != RG_MIGRATE) {
		if (!node_specified)
			printf("Local machine %s %s", actionstr, svcname);
		else
			printf("Member %s %s %s", nodename, actionstr, svcname);
		printf("...");
		fflush(stdout);
		if (msg_open(MSG_SOCKET, 0, RG_PORT, &ctx, 5) < 0) {
			printf("Could not connect to resource group manager\n");
			return 1;
		}
	} else {
		if (!svctarget)
			printf("Trying to %s %s", actionstr, svcname);
		else 
			printf("Trying to %s %s to %s", actionstr, svcname,
			       nodename);
		printf("...");
		fflush(stdout);
		if (msg_open(MSG_SOCKET, 0, RG_PORT, &ctx, 5) < 0) {
			printf("Could not connect to resource group manager\n");
			return 1;
		}
	}

	if (ctx.type < 0) {
		fprintf(stderr,
			"Could not connect to resource group manager!\n");
		return 1;
	}

	if (msg_send(&ctx, &msg, sizeof(msg)) < sizeof(msg)) {
		perror("msg_send");
		fprintf(stderr,
			"Error sending message to resource group manager.\n");
		return 1;
	}

	/* Reusing opt here */
	if ((opt = msg_receive(&ctx, &msg, sizeof(msg), 0)) < sizeof(*h)) {
		perror("msg_receive");
		fprintf(stderr, "Error receiving reply!\n");
		return 1;
	}

	/* Decode */
	if (opt < sizeof(msg)) {
		swab_generic_msg_hdr(h);
		printf("%s\n", rg_strerror(h->gh_arg1));
		return h->gh_arg1;
	}

	swab_SmMessageSt(&msg);
	printf("%s\n", rg_strerror(msg.sm_data.d_ret));
	if (msg.sm_data.d_ret == RG_ERUN)
		return 0;
	if (msg.sm_data.d_ret)
		return msg.sm_data.d_ret;
	
	switch (action) {
	/*case RG_MIGRATE:*/
	case RG_RELOCATE:
	case RG_START:
	case RG_ENABLE:
		printf("%s%s is now running on %s\n",
		       (!node_specified ||
		       msg.sm_data.d_svcOwner==svctarget)?"":"Warning: ",
		       svcname, memb_id_to_name(membership,
		       			        msg.sm_data.d_svcOwner));
		break;
	default:
		break;
	}
	
	return msg.sm_data.d_ret;
}
